### PR TITLE
feat(#9231): add filter in aggregate targets

### LINKED ISSUE #9231: Allow users to view aggregate targets and filter by facility_id and reporting period
**What feature do you want to improve?**
After changes to allow users to have multiple places in [4.9.0](https://docs.communityhealthtoolkit.org/core/releases/4.9.0/) we temporarily disabled users assigned to multiple facilities to view aggregate targets. 

Currently in `4.9.0`, users assigned to multiple `facility_ids` cannot view aggregate targets.  

**Describe the improvement you'd like**
 - Enable users assigned to multiple facilities to view target aggregates page 
 - Update Target Aggregates Component to display filter button
 - Add sidebar filter that is displayed when the filter button is clicked
 - Allow users to filter aggregate targets by reporting period (current month or previous month)
 - Add radio buttons to the sidebar filter that allows users assigned to multiple facilities to filter aggregate targets by facility or reporting period
 -  Add radio buttons to the sidebar filter that allows users assigned to one facility to filter aggregate targets by reporting period  
**Additional context**

screens

### PR DESCRIPTION
**Pull request by Benmuiruri**
# Description

This PR combines 5 PRs to allow: 
- Filtering Aggregate Targets based on facility and reporting period for users assigned to multiple areas.
- Filtering Aggregate Targets based on reporting period for users assigned to one area.

The PRs being combined are: 
1. [Enable place filter aggregate targets for users with multiple](https://github.com/medic/cht-core/pull/9232)
2. [Filter target aggregate data using sidebar filter](https://github.com/medic/cht-core/pull/9267)
3. [Add facility indicator in aggregates targets](https://github.com/medic/cht-core/pull/9282)
4. [Filter aggregate targets by period](https://github.com/medic/cht-core/pull/9283)
5. [e2e tests for aggregate targets (place, period)](https://github.com/medic/cht-core/pull/9305)

For details about what each PR does, please check the description of the individual PRs. 

Closes #9231 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9231-add-aggregate-filter/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9231-add-aggregate-filter/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9231-add-aggregate-filter/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/services/analytics-modules.service.ts [hunk 1]
```diff
@@ -3,6 +3,8 @@ import { Injectable } from '@angular/core';
 import { SettingsService } from '@mm-services/settings.service';
 import { TargetAggregatesService } from '@mm-services/target-aggregates.service';
 
+export const AGGREGATE_TARGETS_ID: string = 'target-aggregates';
```
Comment (Benmuiruri) [inline webapp/src/ts/services/analytics-modules.service.ts:6]: Inferred type

```suggestion
export const AGGREGATE_TARGETS_ID = 'target-aggregates';
```

**Review by Benmuiruri — COMMENTED** [review]
I tested these scenarios 

- Does undefined break anything ? ✅
- If 2025 jan check previous ✅
- Check leap year. ✅
- Confirm admin. ✅
- Online vs offline users ✅
- place with very long name ✅

I didn't find anything apart from those two small code suggestions.

**Comment by latin-panda** [comment]
I tested:
- Reports filter should work and look the same as we had before ✅
- User with one facility assigned - the facility_id is a string (backward compatibility) ✅
- User with two facilities assigned ✅
- Aggregate targets should reset the filter and work fine if there is an app update and the user reloads the app. ✅
- Aggregate targets should reset the filter and work fine if the user clicks a back button or navigates back and forth from one page to another.  ✅
- Admin user gets expected error when trying to access aggregate targets ✅
- Mobile view and Desktop view looking according to Figma  ✅
- We have a test server with a config from a partner and it's working. ✅

File: webapp/src/ts/services/target-aggregates.service.ts [hunk 2]
```diff
@@ -233,23 +255,28 @@ export class TargetAggregatesService {
     if (!facility_id?.length) {
```
[the remark this answers is not in the document]
Comment (Benmuiruri) [reply webapp/src/ts/services/target-aggregates.service.ts:255]: That's a good idea but I would suggest we just leave it as getUserFacilityIds. 
- If we change to using the new this.userSettingsService.getUserFacilities, we have to map the result to extract the id from the facility objects.... we just need id, not whole facility
- We will need to update all tests in `webapp/tests/karma/ts/services/target-aggregates.service.spec.ts` that use `userSettingsService.get`
- We still have `userSettingsService.get` as an existing function even if we dont use it here. 

Therefore for the scope of this task I would say lets leave it.

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/services/target-aggregates.service.ts:255]: Sorry, I didn't understand this one:

> We still have userSettingsService.get as an existing function even if we dont use it here.

This is the only place we use it in this file. Can you please explain a bit more what do you mean by `We still have userSettingsService.get as an existing function`?

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/services/target-aggregates.service.ts:255]: I mean changing this function to this, which is 1 line
```
const facility_ids: string[] = (await this.userSettingsService. getUserFacilities()).map(facility => facility._id);
```

[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/services/target-aggregates.service.ts:255]: This comment is about updating just this file `target-aggregates.service.ts`, the `getUserFacilityIds()` function only. We don't need to update other files.

**Review by tatilepizs — COMMENTED** [review]
I was doing some exploratory testing around Reports, Contacts, and Targets, and two things caught my attention.

1. If I have a place with a really long name and I apply the period filter, the month is not going to be shown in list view. Is that expected? Wouldn't be a better idea to cut the place name in way that the last month is always shown no matter how big the name of place is?

[video](https://github.com/user-attachments/assets/1ee28416-60f2-48ef-b432-8037f3039f55)

2. I created a user with 3 places, at the area level, and everything is working fine, but in the reports tab, the place filter is showing the health facility level as `unavailable`

image

Everything else looks perfectly 😊

File: tests/e2e/default/targets/target-aggregates.wdio-spec.js [hunk 3]
```diff
@@ -109,317 +49,309 @@ describe('Target aggregates', () => {
...
       await commonPage.logout();
       await utils.revertDb(DOCS_TO_KEEP, true);
     });
 
-    it('should disable content when user has many facilities associated', async () => {
-      await loginPage.login({ password: userWithManyPlacesPass, username: userWithManyPlaces.name });
-      await commonPage.waitForPageLoaded();
-      await targetAggregatesPage.checkContentDisabled();
-    });
+    describe('Online user with one place associated', () => {
```
Comment (lorerod) [inline tests/e2e/default/targets/target-aggregates.wdio-spec.js:125]: Is it the same behavior for an `offline` user with only one place associated?
If it is not, we should add the scenario.

File: tests/e2e/default/targets/target-aggregates.wdio-spec.js [hunk 4]
```diff
@@ -109,317 +49,309 @@ describe('Target aggregates', () => {
...
+        await browser.back();
+        const secondTargetItem =
+          await (await targetAggregatesPage.getTargetItem(expectedTargets[1], CURRENT_PERIOD, districtHospital1.name));
+        await helperFunctions.assertTitle(secondTargetItem.title, expectedTargets[1].title);
+      });
+    });
 
-      await commonPage.goToAnalytics();
-      await targetAggregatesPage.goToTargetAggregates(true);
+    describe('Offline user with multiple places associated', () => {
```
Comment (lorerod) [inline tests/e2e/default/targets/target-aggregates.wdio-spec.js:290]: Same question here. Is it the same behavior for `online` users with multiple places associated?

**Review by lorerod — COMMENTED** [review]
I only review the e2e test part of the code. Great work, @tatilepizs. Thank you for completing this work. I left some comments related to the scenarios covered.

File: tests/e2e/default/targets/target-aggregates.wdio-spec.js (hunk 3 above)
Comment (tatilepizs) [reply tests/e2e/default/targets/target-aggregates.wdio-spec.js:125]: Yes it is. The different behavior depends on whether the user has one or more than one place associated.

**Comment by latin-panda** [comment]
@tatilepizs Thanks for exploring the feature more! 

> I created a user with 3 places, at the area level, and everything is working fine, but in the reports tab, the place filter is showing the health facility level as unavailable

CHT expects the facility to have a `name` property in the doc ([see the condition in code](https://github.com/medic/cht-core/blob/f40feb1d9ac2894ec138f2ec887d839f19e2270d/webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts#L240)). When it doesn't have it and user is online, CHT displays `[deleted]`. When the user is offline, it displays `[unavailable]`. I guess this is to handle edge cases, to allow users to keep using the filter (like selecting descendants or seeing what's associated with that unknown place?) 

This is a component only used in Reports page.

File: tests/e2e/default/targets/target-aggregates.wdio-spec.js (hunk 3 above)
Comment (latin-panda) [reply tests/e2e/default/targets/target-aggregates.wdio-spec.js:125]: Correct. It's based on associated places to a user

Comment (Benmuiruri) [reply tests/e2e/default/targets/target-aggregates.wdio-spec.js:125]: Yes Tatiana you are correct. The difference in behavior is whether user has one place or more than one place.

File: tests/e2e/default/targets/target-aggregates.wdio-spec.js (hunk 4 above)
Comment (Benmuiruri) [reply tests/e2e/default/targets/target-aggregates.wdio-spec.js:290]: Yes, you are correct. It is the same behavior for online and offline users

**Comment by latin-panda** [comment]
> If I have a place with a really long name and I apply the period filter, the month is not going to be shown in the list view. Is that expected? Wouldn't it be a better idea to cut the place name in a way that the last month is always shown no matter how big the name of the place is?

@tatilepizs This list uses the "breadcrumbs", the same component as Reports, Contacts, and Tasks. It was designed to show 2 lines of text, then cut it with an ellipsis, so small screens can fit more items in the list (instead of pushing down with long text). 
I tried to add the ellipsis in each section of the breadcrumb, but, users will see less text. It can't go in 2 lines and put the ellipsis at the same time because the css `display` conflicts. 

Screenshot 2024-08-14 at 12 31 43 PM
<br>

We prefer to leave it as it is today, because, it fits more text (reasonably long names) and the reporting period still fits. So,  this as an edge case. 

Screenshot 2024-08-14 at 12 57 22 PM

<br>
Users can still click on a target, and see the reporting period in the details view. 
Screenshot 2024-08-14 at 12 59 28 PM

**Review by latin-panda — APPROVED** [review]
@Benmuiruri this looks good to me!

**Comment by Benmuiruri** [comment]
Thanks everyone for chipping in to get this over the line.
